src/devices/bus/hp_dio/hp98543.cpp

```diff
@@ -0,0 +1,159 @@
+// license:BSD-3-Clause
+// copyright-holders:Sven Schnelle
+/***************************************************************************
+
+  HP98543 medium-resolution color board
+
+***************************************************************************/
+
+#include "emu.h"
+#include "hp98543.h"
+#include "screen.h"
+#include "video/topcat.h"
+#include "video/nereid.h"
+
+#define HP98543_SCREEN_NAME   "98543_screen"
+#define HP98543_ROM_REGION    "98543_rom"
+
+ROM_START(hp98543)
+	ROM_REGION(0x2000, HP98543_ROM_REGION, 0)
+	ROM_LOAD("1818-3907.bin", 0x000000, 0x002000, CRC(5e2bf02a) SHA1(9ba9391cf39624ef8027ce42c84e100344b2a2b8))
+ROM_END
+
+DEFINE_DEVICE_TYPE(HPDIO_98543, dio16_98543_device, "dio98543", "HP98543 medium-res color DIO video card")
+
+MACHINE_CONFIG_START(dio16_98543_device::device_add_mconfig)
+	MCFG_SCREEN_ADD(HP98543_SCREEN_NAME, RASTER)
+	MCFG_SCREEN_UPDATE_DEVICE(DEVICE_SELF, dio16_98543_device, screen_update)
+	MCFG_SCREEN_SIZE(1024,400)
+	MCFG_SCREEN_VISIBLE_AREA(0, 1024-1, 0, 400-1)
+	MCFG_SCREEN_REFRESH_RATE(70)
+
+	MCFG_DEVICE_ADD("topcat_plane_0", TOPCAT, XTAL(35904000))
+	MCFG_TOPCAT_FB_WIDTH(1024)
+	MCFG_TOPCAT_FB_HEIGHT(400)
+	MCFG_TOPCAT_PLANEMASK(1)
+	MCFG_TOPCAT_VRAM(&m_vram)
+
+	MCFG_DEVICE_ADD("topcat_plane_1", TOPCAT, XTAL(35904000))
+	MCFG_TOPCAT_FB_WIDTH(1024)
+	MCFG_TOPCAT_FB_HEIGHT(400)
+	MCFG_TOPCAT_PLANEMASK(2)
+	MCFG_TOPCAT_VRAM(&m_vram)
+
+	MCFG_DEVICE_ADD("topcat_plane_2", TOPCAT, XTAL(35904000))
+	MCFG_TOPCAT_FB_WIDTH(1024)
+	MCFG_TOPCAT_FB_HEIGHT(400)
+	MCFG_TOPCAT_PLANEMASK(4)
+	MCFG_TOPCAT_VRAM(&m_vram)
+
+	MCFG_DEVICE_ADD("topcat_plane_3", TOPCAT, XTAL(35904000))
+	MCFG_TOPCAT_FB_WIDTH(1024)
+	MCFG_TOPCAT_FB_HEIGHT(400)
+	MCFG_TOPCAT_PLANEMASK(8)
+	MCFG_TOPCAT_VRAM(&m_vram)
+
+	MCFG_DEVICE_ADD("nereid", NEREID, 0)
+MACHINE_CONFIG_END
+
+const tiny_rom_entry *dio16_98543_device::device_rom_region() const
+{
+	return ROM_NAME(hp98543);
+}
+
+dio16_98543_device::dio16_98543_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	dio16_98543_device(mconfig, HPDIO_98543, tag, owner, clock)
+{
+}
+
+dio16_98543_device::dio16_98543_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, type, tag, owner, clock),
+	device_dio16_card_interface(mconfig, *this),
+	m_topcat0(*this, "topcat_plane_0"),
+	m_topcat1(*this, "topcat_plane_1"),
+	m_topcat2(*this, "topcat_plane_2"),
+	m_topcat3(*this, "topcat_plane_3"),
+	m_nereid(*this, "nereid")
+{
+	m_vram.resize(0x80000);
```

Comment:
Why are you using a vector for this if it's fixed size?  Why are you allocating at construction and not deferring until start?